velox/functions/prestosql/ArrayDuplicates.cpp

```diff
@@ -0,0 +1,217 @@
+      exec::Expr* caller,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    // Acquire the array elements vector.
+    auto arrayVector = args.front()->as<ArrayVector>();
+    auto elementsVector = arrayVector->elements();
+    ArrayDuplicatesFunction::validateElementType(elementsVector->typeKind());
+
+    auto elementsRows =
+        toElementRows(elementsVector->size(), rows, arrayVector);
+    exec::LocalDecodedVector elements(context, *elementsVector, elementsRows);
+
+    vector_size_t elementsCount = elementsRows.size();
+    vector_size_t rowCount = arrayVector->size();
+
+    // Allocate new vectors for indices, length and offsets.
+    memory::MemoryPool* pool = context->pool();
+    BufferPtr newIndices =
+        AlignedBuffer::allocate<vector_size_t>(elementsCount, pool);
+    BufferPtr newLengths =
+        AlignedBuffer::allocate<vector_size_t>(rowCount, pool);
+    BufferPtr newOffsets =
+        AlignedBuffer::allocate<vector_size_t>(rowCount, pool);
+
+    // Pointers and cursors to the raw data.
+    vector_size_t indicesCursor = 0;
+    auto* rawNewIndices = newIndices->asMutable<vector_size_t>();
+    auto* rawSizes = newLengths->asMutable<vector_size_t>();
+    auto* rawOffsets = newOffsets->asMutable<vector_size_t>();
+
+    // Process the rows: use a hashmap to store unique values and
+    // whether it occurred once or more than once.
+    folly::F14FastMap<T, bool> uniqueMap;
+
+    rows.applyToSelected([&](vector_size_t row) {
+      auto size = arrayVector->sizeAt(row);
+      auto offset = arrayVector->offsetAt(row);
+
+      rawOffsets[row] = indicesCursor;
+      vector_size_t numOfNulls = 0;
+      T prevDuplicateValue; // [Sort] for compare and skip-null purpose
+
+      for (vector_size_t i = offset; i < offset + size; ++i) {
+        if (elements->isNullAt(i)) {
+          numOfNulls++;
+          if (numOfNulls == 2) {
+            rawNewIndices[indicesCursor] = i;
+            // [Sort] prepend the null to head of the array
+            if (indicesCursor > rawOffsets[row]) {
+              vector_size_t nullNewIdx = rawNewIndices[indicesCursor];
+              for (vector_size_t j = indicesCursor; j > rawOffsets[row]; --j) {
+                rawNewIndices[j] = rawNewIndices[j - 1];
+              }
+              rawNewIndices[rawOffsets[row]] = nullNewIdx;
+            }
+
+            indicesCursor++;
+          }
+        } else {
+          T value = elements->valueAt<T>(i);
+          auto it = uniqueMap.find(value);
+          if (it == uniqueMap.end()) {
+            uniqueMap[value] = true;
+          } else if (it->second) {
+            it->second = false;
+            rawNewIndices[indicesCursor] = i;
+            // [Sort] if it is the first non-null element, do nothing;
+            // otherwise, swap with the previous non-null element if the current
+            // is less
+            if (!((indicesCursor == rawOffsets[row] + 1 && numOfNulls > 1) ||
```

Comment:
@mbasmanova I understand the sorting logic is a bit hard to understand even with inline comments.
Yes, this new logic passed the following two unit test cases:


I also added a section "Sorting Algorithm Explanation" in this PR to explain more details. The time complexity will be O(N) because of only one-time scan, which is better than std::sort() (O(N*Log(N)) on the duplicate result array.

Reply:
I wanna achieve sorting in the original one-time scan. Looks like I missed one corner case when insert a non-null value. Let me revise the logic.

Reply:
Take my word back, reimplement it with std::sort plus a nullable sort comparator.